src/apps/voip/VoipPacketSerializer.cc

```diff
@@ -0,0 +1,62 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see http://www.gnu.org/licenses/.
+//
+
+#include "VoipPacket_m.h"
+#include "apps/voip/VoipPacketSerializer.h"
+#include "inet/common/packet/serializer/ChunkSerializerRegistry.h"
+
+using namespace inet;
+
+Register_Serializer(VoipPacket, VoipPacketSerializer);
+
+void VoipPacketSerializer::serialize(MemoryOutputStream& stream, const Ptr<const Chunk>& chunk) const
+{
+    auto startPosition = stream.getLength();
+    const auto& voipPacket = staticPtrCast<const VoipPacket>(chunk);
+    stream.writeUint32Be(B(voipPacket->getChunkLength()).get());
+    stream.writeUint32Be(voipPacket->getIDtalk());
+    stream.writeUint32Be(voipPacket->getNframes());
+    stream.writeUint32Be(voipPacket->getIDframe());
+    stream.writeUint64Be(voipPacket->getTimestamp().raw());
+    stream.writeUint64Be(voipPacket->getArrivalTime().raw());
+    stream.writeUint64Be(voipPacket->getPlayoutTime().raw());
+    stream.writeUint32Be(voipPacket->getSize());
+
+    int64_t remainders = B(voipPacket->getChunkLength() - (stream.getLength() - startPosition)).get();
+    if (remainders < 0)
+        throw cRuntimeError("VoipPacket length = %d smaller than required %d bytes", (int)B(voipPacket->getChunkLength()).get(), (int)B(stream.getLength() - startPosition).get());
```

Comment:
The next INET version will automatically check if the serialized and non-serialized forms have the exact same length.

Reply:
You could also use a separate chunk for the data part.

Reply:
For the time being, I left the size check - we can remove it, when the next INET version is available. A separate data chunk would be an option, but right now we would not really benefit.